frame/democracy/src/lib.rs

```diff
@@ -1676,19 +1932,117 @@ impl<T: Trait> Module<T> {
 	}
 
 	/// Current era is ending; we should finish up any proposals.
-	fn begin_block(now: T::BlockNumber) -> DispatchResult {
+	///
+	///
+	/// # <weight>
+	/// If a referendum is launched or maturing take full block weight. Otherwise:
+	/// - Complexity: `O(R)` where `R` is the number of unbaked referenda.
+	/// - Db reads: `LastTabledWasExternal`, `NextExternal`, `PublicProps`, `account`,
+	///   `ReferendumCount`, `LowestUnbaked`
+	/// - Db writes: `PublicProps`, `account`, `ReferendumCount`, `DepositOf`, `ReferendumInfoOf`
+	/// - Db reads per R: `DepositOf`, `ReferendumInfoOf`
+	/// # </weight>
+	fn begin_block(now: T::BlockNumber) -> Result<Weight, DispatchError> {
+		let mut weight = 60_000_000 + T::DbWeight::get().reads_writes(6, 5);
+
 		// pick out another public referendum if it's time.
 		if (now % T::LaunchPeriod::get()).is_zero() {
 			// Errors come from the queue being empty. we don't really care about that, and even if
 			// we did, there is nothing we can do here.
 			let _ = Self::launch_next(now);
+			weight = T::MaximumBlockWeight::get();
 		}
 
 		// tally up votes for any expiring referenda.
-		for (index, info) in Self::maturing_referenda_at(now).into_iter() {
+		let next = Self::lowest_unbaked();
+		let last = Self::referendum_count();
+		let r = Weight::from(last.saturating_sub(next));
+		weight += 11_000_000 * r + T::DbWeight::get().reads(2 * r);
+		for (index, info) in Self::maturing_referenda_at_inner(now, next..last).into_iter() {
 			let approved = Self::bake_referendum(now, index, info)?;
 			ReferendumInfoOf::<T>::insert(index, ReferendumInfo::Finished { end: now, approved });
+			weight = T::MaximumBlockWeight::get();
+		}
+
+		Ok(weight)
+	}
+
+	/// Reads the length of PublicProps without getting the complete value in the runtime.
+	fn len_of_public_props() -> u32 {
+		decode_compact_u32_at(&<PublicProps<T>>::hashed_key()).unwrap_or(0)
+	}
+
+	/// Reads the length of account in DepositOf without getting the complete value in the runtime.
+	///
+	/// Return 0 if no deposit for this proposal.
+	fn len_of_deposit_of(proposal: PropIndex) -> Option<u32> {
+		decode_compact_u32_at(&<DepositOf<T>>::hashed_key_for(proposal))
+	}
+
+	/// Check pre image is missing variant without getting the complete value in the runtime.
+	fn check_pre_image_is_missing(proposal_hash: T::Hash) -> DispatchResult {
+		let mut buf = [0u8; 1];
+		let key = <Preimages<T>>::hashed_key_for(proposal_hash);
+		let bytes = match sp_io::storage::read(&key, &mut buf, 0) {
+			Some(bytes) => bytes,
+			None => return Err(Error::<T>::NotImminent.into()),
+		};
+		let mut input = &buf[0..buf.len().min(bytes as usize)];
+
+		match input.read_byte() {
+			Ok(0) => Ok(()),
+			Ok(1) => Err(Error::<T>::DuplicatePreimage.into()),
+			_ => {
+				sp_runtime::print("Failed to decode `PreimageStatus` variant");
+				Err(Error::<T>::NotImminent.into())
+			}
+		}
+	}
```

Comment:
This is a more efficient `exists`?

Could you add some comments about what is going on here?

Reply:
I think democracy pre_image values can be very big (with runtime upgrade for example).

This functions checks that the value is PreImageStatus::Missing variant without getting the raw value in the runtime only the first byte.

hmm function name is a bit confusing. but this doesn't just check that value exists.

Reply:
So if the `PreImageStatus` enum changes or is removed, or the storage item is changed, this function would need to change.

I assume the test you have placed will confirm that this storage item and ultimately this function does NOT change in any way we dont expect?

Reply:
yes tests are there https://github.com/paritytech/substrate/pull/5828/files#diff-f94cd6599f816b386fc6b68ef980e8c3